merger.go

```diff
@@ -0,0 +1,94 @@
+package main
+
+import (
+	"go/ast"
+	"go/token"
+
+	"golang.org/x/tools/go/packages"
+)
+
+// MergePackageFiles creates a file AST by merging the ASTs of the
+// files belonging to a package. The mode flags control merging behavior.
+func MergePackageFiles(pkg *packages.Package) *ast.File {
+	// Count the number of package docs, comments and declarations across
+	// all package files. Also, compute sorted list of filenames, so that
+	// subsequent iterations can always iterate in the same order.
+	var minPos, maxPos token.Pos
+	i := 0
+	var pos token.Pos
+	var ndecls = 0
+	for _, f := range pkg.Syntax {
+		pkg.Name = f.Name.Name
+		i++
+		ndecls += len(f.Decls)
+		if i == 0 || f.FileStart < minPos {
+			minPos = f.FileStart
+		}
+		if i == 0 || f.FileEnd > maxPos {
+			maxPos = f.FileEnd
+		}
+		if f.Package > pos {
+			// Keep the maximum package clause position as
+			// position for the package clause of the merged
+			// files.
+			pos = f.Package
+		}
+	}
+
+	// Collect import specs from all package files.
+	var imports = []*ast.ImportSpec{}
+	for _, f := range pkg.Syntax {
+		for _, imp := range f.Imports {
+			notDupe := true
+			for _, imp2 := range imports {
+				if imp.Path.Value == imp2.Path.Value {
+					notDupe = false
+				}
+			}
+			if notDupe {
+				imports = append(imports, imp)
+			}
+		}
+	}
```

Comment:
Use a map as a set (also requires changes later on in the code):
```suggestion
	// Collect import specs from all package files.
	imports := make(map[*ast.ImportSpec]struct{})
	for _, f := range pkg.Syntax {
		for _, imp := range f.Imports {
				imports[imp] = struct{}{}
		}
	}
```